File: code_snippets/ts/src/develop/workflows/signup.ts
import userManagement from "./service";
// <start_here>
import * as restate from "@restatedev/restate-sdk";
import {WorkflowContext} from "@restatedev/restate-sdk";

const signUpWorkflow = restate.workflow({
    name: "signup",
    handlers: {
        // <mark_1>
        run: async (ctx: WorkflowContext, req: { email: string }) => {
            const secret = ctx.rand.uuidv4();
            ctx.set("status", "Generated secret");

            await ctx.run("send email", () =>
                sendEmailWithLink({ email: req.email, secret }));
            ctx.set("status", "Sent email");

            // <mark_3>
            const clickSecret = await ctx.promise<string>("email.clicked");
            // </mark_3>
            ctx.set("status", "Clicked email");

            return clickSecret == secret;
        },
        // </mark_1>


        click: (ctx: restate.WorkflowSharedContext, secret: string) =>
            // <mark_3>
            ctx.promise<string>("email.clicked").resolve(secret),
        // </mark_3>

        // <mark_2>
        getStatus: (ctx: restate.WorkflowSharedContext) =>
            ctx.get<string>("status"),
        // </mark_2>
    },
});

export type SignUpWorkflow = typeof signUpWorkflow;

// <mark_4>
restate.endpoint().bind(signUpWorkflow).listen();
// </mark_4>
// <end_here>

function sendEmailWithLink(param: { email: string, secret: string}){
    console.log(`Sending email to ${param.email} with secret ${param.secret}`);
}

File: code_snippets/python/src/develop/workflows/signup.py
from typing import Optional
import uuid

import restate
from restate import Workflow, WorkflowContext, WorkflowSharedContext
from src.develop.workflows.email_client import EmailClient


# <start_here>
signup_workflow = Workflow("SignupWorkflow")
email_client = EmailClient()


# <mark_1>
@signup_workflow.main()
async def run(ctx: WorkflowContext, email: str):
    secret = await ctx.run("secret", lambda: str(uuid.uuid4()))
    ctx.set("onboarding_status", "Generated secret")

    await ctx.run("send email", lambda: email_client.send_email_with_link(email, secret))
    ctx.set("onboarding_status", "Sent email")

    # <mark_3>
    click_secret = await ctx.promise("email.clicked", type_hint=str).value()
    # </mark_3>
    ctx.set("onboarding_status", "Clicked email")

    return click_secret == secret


# </mark_1>


@signup_workflow.handler()
async def click(ctx: WorkflowSharedContext, secret: str):
    # <mark_3>
    await ctx.promise("email.clicked", type_hint=str).resolve(secret)
    # </mark_3>


# <mark_2>
@signup_workflow.handler()
async def get_status(ctx: WorkflowSharedContext) -> str:
    return await ctx.get("onboarding_status") or "Not started yet"


# </mark_2>

# <mark_4>
app = restate.app([signup_workflow])
# </mark_4>
# <end_here>
